src/webgpu/shader/execution/evaluation_order.spec.ts

```diff
@@ -0,0 +1,480 @@
+export const description = `
+Execution Tests for evaluation order of expressions
+`;
+
+import { makeTestGroup } from '../../../common/framework/test_group.js';
+import { GPUTest } from '../../gpu_test.js';
+import { TypeI32 } from '../../util/conversion.js';
+
+export const g = makeTestGroup(GPUTest);
+
+const common_source = `
+var<private> a : i32 = 2;
+var<private> b : i32 = 3;
+var<private> c : i32 = 4;
+
+var<private> arr2D : array<array<i32, 10>, 10> = array<array<i32, 10>, 10>(
+  array<i32, 10>( 0,  1,  2,  3,  4,  5,  6,  7,  8,  9),
+  array<i32, 10>(10, 11, 12, 13, 14, 15, 16, 17, 18, 19),
+  array<i32, 10>(20, 22, 22, 23, 24, 25, 26, 27, 28, 29),
+  array<i32, 10>(30, 32, 32, 33, 34, 35, 36, 37, 38, 33),
+  array<i32, 10>(40, 42, 42, 43, 44, 45, 46, 47, 48, 44),
+  array<i32, 10>(50, 52, 52, 53, 54, 55, 56, 57, 58, 55),
+  array<i32, 10>(60, 62, 62, 63, 64, 65, 66, 67, 68, 66),
+  array<i32, 10>(70, 72, 72, 73, 74, 75, 76, 77, 78, 77),
+  array<i32, 10>(80, 82, 82, 83, 84, 85, 86, 87, 88, 98),
+  array<i32, 10>(90, 92, 92, 93, 94, 95, 96, 97, 98, 99),
+);
+
+var<private> arr1D_zero : array<i32, 50>;
+var<private> arr2D_zero : array<array<i32, 50>, 50>;
+var<private> arr3D_zero : array<array<array<i32, 50>, 50>, 50>;
+
+var<private> vec4_zero : vec4<i32>;
+
+struct S {
+  x : i32;
+  y : i32;
+  z : i32;
+}
+
+fn mul(p1 : ptr<private, i32>, multiplicand : i32) -> i32 {
+  *p1 = *p1 * multiplicand;
+  return *p1;
+}
+
+fn add(p1 : ptr<private, i32>, addend : i32) -> i32 {
+  *p1 = *p1 + addend;
+  return *p1;
+}
+
+fn sub3(a : i32, b : i32, c : i32) -> i32 {
+  return a - b - c;
+}
+
+fn bool_to_i32(expr : bool) -> i32 {
+  if (expr) {
+    return 1;
+  }
+  return 0;
+}
+
+fn make_vec3(init : i32) -> vec3<i32> {
+  return vec3<i32>(init, init, init);
+}
```

Comment:
The type constructor `vec3(i)` should do this (and in fewer characters), so not sure you need this either.

Reply:
Done.